helper/fairshare/jobmanager.go

```diff
@@ -203,57 +285,48 @@ func (j *JobManager) assignWork() {
 				default:
 				}
 
-				job := j.getNextJob()
+				job, queueID := j.getNextJob()
 				if job != nil {
-					j.workerPool.dispatch(job)
+					j.workerPool.dispatch(job,
+						func() {
+							j.incrementWorkerCount(queueID)
+						},
+						func() {
+							j.decrementWorkerCount(queueID)
+						})
 				} else {
 					break
 				}
 			}
 
-			// listen for a wake-up when an emtpy job manager has been given
-			// new work
 			select {
 			case <-j.quit:
 				j.wg.Done()
 				return
 			case <-j.newWork:
-				break
+				// listen for wake-up when an empty job manager has been given work
+			case <-time.After(50 * time.Millisecond):
+				// periodically check if new workers can be assigned. with the
+				// fairsharing worker distribution it can be the case that there
+				// is work waiting, but no queues are elligible for another worker
 			}
 		}
 	}()
 }
 
 // addQueue generates a new queue if a queue for `queueID` doesn't exist
+// it also starts tracking workers on that queue, if not already tracked
 // note: this must be called with l held for write
 func (j *JobManager) addQueue(queueID string) {
 	if _, ok := j.queues[queueID]; !ok {
 		j.queues[queueID] = list.New()
-		j.queuesIndex = append(j.queuesIndex, queueID)
 	}
-}
-
-// removeLastQueueAccessed removes the queue and index map for the last queue
-// accessed. It is to be used when the last queue accessed has emptied.
-// note: this must be called with l held for write
-func (j *JobManager) removeLastQueueAccessed() {
-	if j.lastQueueAccessed == -1 || j.lastQueueAccessed > len(j.queuesIndex)-1 {
-		j.logger.Warn("call to remove queue out of bounds", "idx", j.lastQueueAccessed)
-		return
-	}
-
-	queueID := j.queuesIndex[j.lastQueueAccessed]
-
-	// remove the queue
-	delete(j.queues, queueID)
-
-	// remove the index for the queue
-	j.queuesIndex = append(j.queuesIndex[:j.lastQueueAccessed], j.queuesIndex[j.lastQueueAccessed+1:]...)
 
-	// correct the last queue accessed for round robining
-	if j.lastQueueAccessed > 0 {
-		j.lastQueueAccessed--
-	} else {
-		j.lastQueueAccessed = len(j.queuesIndex) - 1
+	// it's possible the queue ran out of work and was pruned, but there were
+	// still workers operating on data formerly in that queue, which were still
+	// being tracked. if that is the case, we don't want to wipe out that worker
+	// count when the queue is re-initialized.
+	if _, ok := j.workerCount[queueID]; !ok {
+		j.workerCount[queueID] = 0
```

Comment:
I'm not sure I understand this part -- if there are still operating workers in the queue, isn't their count still at 0?

Reply:
if there are 5 workers operating on data from a particular queue, the count of how many workers will be 5 (and that will decrease as workers finish their tasks). the edge case is if there is no upcoming work in that queue, but workers are still operating on stuff that came from that queue

Reply:
Oh, I see. So is the reason this is being explicitly set to 0 in this case because `DecrementWorkers` starts by decreasing the workerCount, and then checks `j.workerCount[queueID]` ?

Reply:
well, this is more about only re-initializing if we aren't already tracking data there. i think the more important bit is the if statement - setting to 0 just initializes the `queueID` k/v in the map (but we have to be careful to not wipe out data if it's already there)

does that make sense? is there a way i could improve the comment to make it more clear?

Reply:
Yeah, that makes sense! I'm not sure about the comment reformatting, but in light of this conversation the comment looks good.